src/extension.ts

```diff
@@ -279,31 +312,45 @@ class ClientWorkspace {
     }
 
     this.disposables.forEach(d => d.dispose());
+    commandsUnregistered = true;
   }
 
-  private registerCommands(context: ExtensionContext) {
+  private registerCommands(
+    context: ExtensionContext,
+    multiProjectEnabled: boolean,
+  ) {
     if (!this.lc) {
       return;
     }
+    if (multiProjectEnabled && !commandsUnregistered) {
+      return;
+    }
 
+    commandsUnregistered = false;
     const rustupUpdateDisposable = commands.registerCommand(
       'rls.update',
       () => {
-        return rustupUpdate(this.config.rustupConfig());
+        const ws =
+          multiProjectEnabled && activeWorkspace ? activeWorkspace : this;
+        return rustupUpdate(ws.config.rustupConfig());
       },
     );
     this.disposables.push(rustupUpdateDisposable);
 
     const restartServer = commands.registerCommand('rls.restart', async () => {
-      await this.stop();
-      return this.start(context);
+      const ws =
+        multiProjectEnabled && activeWorkspace ? activeWorkspace : this;
+      await ws.stop();
+      return ws.start(context);
     });
     this.disposables.push(restartServer);
 
     this.disposables.push(
-      commands.registerCommand('rls.run', (cmd: Execution) =>
-        runRlsCommand(this.folder, cmd),
-      ),
+      commands.registerCommand('rls.run', (cmd: Execution) => {
+        const ws =
```

Comment:
Reading this took me some second to understand that ws was workspace, also personal preference but I find ternary easier to read like so :

```typescript
const workspace = (multiProjectEnabled && activeWorkspace)
      ? activeWorkspace
      : this;
runRlsCommand(workspace.folder, cmd);
```

I don't know if it's ok with the tslint configs.